hw/ip/csrng/doc/_index.md

```diff
@@ -0,0 +1,89 @@
+---
+title: "CSRNG HWIP Technical Specification"
+---
+
+# Overview
+
+This document specifies the Cryptographically Secure Random Number Generator (CSRNG) hardware IP functionality.
+Due to the importance of secure random number generation (RNG), it is a topic which is extensively covered in security standards.
+This IP targets compliance with both [BSI’s AIS31 recommendations for Common Criteria](https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Zertifizierung/Interpretationen/AIS_31_Functionality_classes_for_random_number_generators_e.pdf), as well as [NIST's SP 800-90](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf) (as referenced in [FIPS 140-3](https://csrc.nist.gov/publications/detail/fips/140/3/final)).
+Since these two standards use significantly different terminology, it is recommended that the reader refer to our RNG compliance strategy document for an overview of the various RNG classes and equivalencies between the two standards.
+The CSRNG IP supports both of these standards for both deterministic (DRNG) and true random number generation (TRNG).
+In NIST terms, it works with the entropy_source IP to satisfy the requirements as a DRBG or NRBG (non-deterministic random bit generator).
+In AIS31 language, this same implementation satisfies both the DRG.3 requirements for deterministic generation, or the PTG.3 requirements for cryptographically processes physical generation.
+This IP block is attached to the chip interconnect bus as a peripheral module conforming to the [Comportable guideline for peripheral functionality]({{< relref "doc/rm/comportability_specification" >}}), but also has direct hardware links to other IPs for secure internal transmission of random numbers.
+
+## Features
+- Together with a secure entropy source, provides support for both deterministic (DRNG) and true random number generation (TRNG).
+- Compliant with NIST and BSI recommendations for random number generation
+- In deterministic mode, meets the requirements given in AIS31 for a DRG.3 class deterministic random number generator (DRNG) meaning it provides Forward Secrecy and Enhanced Backward Secrecy.
+- Utilizes the CTR_DRBG construction specified in NIST SP 800-90A, qualifying it as an NIST-approved DRBG (deterministic random bit generator).
+    - Operates at 256bit security strength.
+- Support for multiple separate CSRNG instances per IP block.
+    - Each instance has its own internal state, control, reseed counters and IO pins.
+    - The number of RNG instances is set via a module parameter
+- Software access to a dedicated RNG instance.
+    - One instance, Instance 0, is always accessible through TLUL registers,
+    - All other instances route to other hardware peripherals (e.g. the key manager, obfuscation engines etc.) and in normal operation these other instances are inaccessible from software.
```

Comment:
Did we close on CTR_DRBG or is it still open?
Why do we need more than one RNG instance?
Regardless, I assume there is only going to be one AES instance, right?

Reply:
I believe CTR_DRBG has been the only implementation considered in every conversation I've had.  If there are objections now is probably the right time to voice them though.
We need multiple instances to have protected state for each consumer.  However yes there is only one AES.

Reply:
I would prefer to not hardwire assumptions about there being only 1 AES engine, if at all possible.

As an IP core, I can easily imagine a situation where someone would put want multiple AES cores.  (I realize that's not the first use-case we are targeting, but precluding it would be annoying.)

Reply:
@cdgori as I see the csrng design lay out, the micro-architecture will only use one AES cipher block. Arbiters will be used to select access to the requester(s). Multiple AES block could be used simply by replacing the arbiters with direct connections. This is the trade-off between area, power, latency, etc.

Reply:
Ah, sorry @mwbranstad, I probably have misunderstood the context, I realize now that there are two different AESes we are talking about.

100% agree that there is only the one AES instance in the CSRNG, w/ arbiters as you describe.  I actually can't see needing more than one AES there even in a (speculative) version of OT core with lots of crypto modules, it just seems unlikely to need to generate CSRNG instances with kind of performance.

I thought (wrongly, bias of thinking about the AES client block lately) that this was talking about how many possible AES clients there could be, which was what I was reacting to.